### ISSUE TITLE
Backtesting API gives incorrect results when handling many concurrent requests

### ISSUE THREAD

**Issue by cabloo** [askmike/gekko#1324]
**Note: for support questions, please join our [Discord server](https://discord.gg/26wMygt)**

* **I'm submitting a ...**
[x] bug report
[ ] feature request
[ ] question about the decisions made in the repository

* **Action taken** (what you did)
Generated a lot of concurrent requests to the Gekko Backtesting API with the same query data.

* **Expected result** (what you hoped would happen)
Consistently answer the same query with the same results.

* **Actual result** (unexpected outcome)
Results are wildly different and graphs look completely changed, a lot of data appears to be missing from the graph.

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, etc)
I suspect race conditions with processing so many queries at once?

**Comment by askmike** [comment]
There should not be: each backtest runs in it's own child process.

Could you tell us more about what strategy you are running (using which indicators, and whether they are native or TAlib/Tulip).

**Comment by cabloo** [comment]
Hi Mike,

The strategy was MACD. I'm not home so don't have the exact query I ran but
will send you the full query later tonight.

Thanks,
Zane

**Comment by askmike** [comment]
Strange, if you could provide some more details like: the MACD settings, the backtest timerange and candleSize as well as the market I can try to reproduce.

**Comment by cabloo** [comment]
```json
{
      "gekkoConfig": {
        "watch": {
          "exchange": "kraken",
          "currency": "USD",
          "asset": "BCH"
        },
        "paperTrader": {
          "feeMaker": 0.16,
          "feeTaker": 0.25,
          "feeUsing": "maker",
          "slippage": 0.05,
          "simulationBalance": {
            "asset": 1,
            "currency": 1500
          },
          "reportRoundtrips": true,
          "enabled": true
        },
        "tradingAdvisor": {
          "enabled": true,
          "method": "MACD",
          "candleSize": 63,
          "historySize": 10,
        },
        "MACD":  {
    "short": 3,
    "long": 5,
    "signal": 2,
    "thresholds": {
      "down": -0.0933,
      "up": 0.0379,
      "persistence": 6
    }
  },
        "backtest": {
          "daterange": {
            "from": "2017-11-07T00:41:00Z",
            "to": "2017-11-13T01:31:00Z"
          }
        },
        "performanceAnalyzer": {
          "riskFreeReturn": 2,
          "enabled": true
        },
        "valid": true
      },
      "data": {
        "candleProps": ["close", "start"],
        "indicatorResults": true,
        "report": true,
        "roundtrips": true,
        "trades": true
      }
    };
  }
```

Kraken times:
```
            "from": "2017-11-07T00:41:00Z",
            "to": "2017-11-13T01:31:00Z"
```

I am running 100 requests to the API spaced 100ms apart so about 10 requests per second. It seems to be responding to all requests but the results quickly become incorrect. If I run the backtest using the gekko web UI, at first I get consistent results. Once I start running my script and try the gekko UI backtest again, I get wildly varying results and the graph shows much less granularity. It looks as if the DB is only returning a certain portion of the results.

Script is something like this:
```js
var request = require('request-promise');
var WORKERS = 100;

for (var i = 0; i < WORKERS; i++) {
  setTimeout(runRequest, i*100);
}

function runRequest() {
    return request({
      method: 'POST',
      uri: 'http://localhost:3001/api/backtest',
      body: config,
      json: true,
    }).then(runRequest);
}
```

**Comment by cmroche** [comment]
I can confirm that this is happening. Seems to be any strategy, any pair, and I get differing results with only 10 concurrent processes (first number I tried). If I run a bunch of tests sequentially with the same parameters the results are stable, but as soon as there are concurrent requests things are wrong, in fact none of them are right.

![snip20171122_1](https://user-images.githubusercontent.com/1007595/33155383-59d6fc10-cfbe-11e7-8b2c-ad66bf9db168.png)

**Comment by cmroche** [comment]
More information on this, running on a core i3 process with only 2 cores on the CPU.

When using 2 concurrent processes everything is fine

![snip20171122_2](https://user-images.githubusercontent.com/1007595/33155527-349df65a-cfbf-11e7-8fb3-e46027a6bfcb.png)

When using 3 concurrent processes it breaks

![snip20171122_3](https://user-images.githubusercontent.com/1007595/33155538-4dea504a-cfbf-11e7-8714-05d9b35318d5.png)

I'll validate this now on another processor.

**Comment by cmroche** [comment]
On a 4 core processor I start to get bad results after 2 processes as well.

**Comment by cmroche** [comment]
And... I did timestamp validation of candles in backtest.js (processCandles), data is definitely arriving out of order with concurrent requests.

**Comment by cmroche** [comment]
So as an interesting note, even when the result from the API call is wrong, the output at the end of `PerformanceAnalyzer.prototype.finalize` is always the same (and correct). So it seems gekko is actually calculating the correct profit report at some point.

**Comment by cmroche** [comment]
I've found the problem. In cp.js we're not handling flow control on process.send, the IPC channel is overflowing and messages are being dropped. The fix is to block the send until we can succeed, I'll send a fix shortly.

**Comment by cabloo** [comment]
Awesome debugging work Clifford! Thanks for looking into this.

**Comment by askmike** [comment]
@cmroche you are on fire as of late!

ping @just-ak: I think this is exactly the same issue as you found in #1285.

**Comment by cmroche** [comment]
Turns out the issue was not entirely flow control, nodejs isn't clear in the documentation, but process.send has an internal flow control which does buffer outstanding requests into a queue. The flow control signalling it provides is to enable interrupting external processes like a backpressure system, that isn't necessary in our case and the simple approach will work.

The problem in the end was that the child was being explicitly terminated before the IPC channel could clear the buffer, I re-implemented the child shutdown as an event signalled through the parent, which will enable us to ensure the IPC channel is clear of vital information before terminating.

**Comment by cabloo** [comment]
I believe this issue is still occurring on `develop` branch. 

I checked that my local version has the patch (and restarted docker compose):

```
cabloo@media:~/gekko$ git log
...
commit a871b3edbc75c499b42031d58ac84a21ea6aa9e4
Author: Clifford Roche <clifford.roche@gmail.com>
Date:   Thu Nov 23 20:54:06 2017 -0500

    Fix bug with incorrect backtest results
...
```

While my script that hits the API with 50 requests is running, the backtest results are still inconsistent:
![selection_032](https://user-images.githubusercontent.com/229041/33305920-fe43092e-d3c4-11e7-9dbf-615e51084f45.png)
![selection_031](https://user-images.githubusercontent.com/229041/33305926-01a5676a-d3c5-11e7-9877-a4be6489e932.png)

Note that the start time matches but end time does not match what I selected in either case (I selected 2017-10-25 00:31). Results appear to be truncated at a variable portion of the data.

Here is what the exact same request looks like after I have stopped the script: 
![selection_033](https://user-images.githubusercontent.com/229041/33306004-624dcfda-d3c5-11e7-83d7-bb2aa86ba2d6.png)

My original thought was that the graphs had less granularity but it appears that the actual issue is data truncation while many API requests are being run in parallel. The 2nd screenshot appears to be the same as the first and third screenshot but truncated and then stretched.

**Comment by cmroche** [comment]
I cannot say I have the same issue, it appears to work on my end now. Delete and rebuild the docker images perhaps something is not caching properly there?

Keep in mind that there is a two minute default timeout on the webserver. I'm not sure that this is the source of your issue since I don't thin it leads to the child process being abandoned, but you might want to try increasing it.

**Comment by dzham09** [comment]
The issue does exists, for me i'm testing BTC/USDT entire Feb 2018, using 1 minute candles.  That's about 40k+ ticks/candles.  I printed out the ticks and noticed it changed every time.  There is an async timing issue where it fills some buffer and overflows and probably the strategy can't keep up and the candles are lost.  But after a few attempts the work around is to set the config.backtest.batchSize to 1.  Unfortunately its not consistent, this is still stopping after 5-15 days of data.

**Comment by Jabbaxx** [comment]
I think we need to reopen this or create a new issue.
I am only using talib indicators. No other indicators.
I can confirm this issues also on backtesting. I do a lot of backtesting with gekko and I can see that if I repeat exactly the same test, the results are not the same :(
I have now tried with setting config.backtest.batchSize to 1 and get more consistent results. But I will test some more and see how it goes.

**Comment by Jabbaxx** [comment]
I can confirm that there are big issues. 
If i set 

config.backtest.batchSize to 1

Then there is still some few issues, meaning 1-2 out of 500 backtests are wrong. 
But when I have default 50, it is crazy many. More like 1 out of 10.
I tried simple calculation (sell - Buy)/buy price in the strategy to find the percentage and add that up. It fails also. So it is not round trip related. It must be something in the strategy that goes wrong.

**How can we debug this ?**

**Comment by stowj** [comment]
dzham09 can you please share your gekko version? Node version? Exchange? etc. I’m also experiencing the same issues.

**Comment by dzham09** [comment]
There's nothing special about it, but i've made a lot of changes at one time so its messy and not really compatible with the current version.   Node version i think was 9 something (i've not looked at it for a while). I believe the issue is due to the nature of Node JS is asynchronous.   By narrowing the backSize you force it to process one at a time and its less likely to process them out of order.  Its basically just slowing it down.  Reminds me of the issue people talk about with using Tulip (but I wasn't testing with any external indicators).

**Comment by askmike** [comment]
Guys, I have been working hard on v0.6 which has a completely [rewritten event flow in core](https://forum.gekko.wizb.it/thread-56579.html). It also handles previously unhandled race conditions and has a ton of overall improvements.

Read this post for more info: https://forum.gekko.wizb.it/thread-57279.html

However you guys are doing many concurrent requests presumably via te API, the current 0.6 doesn't have a working API yet. If I were to update this, is anyone able to quickly test whether the problems are still there?

**Comment by Jabbaxx** [comment]
I had same problem. When I changed from using talib to use native
indicators there was no problem anymore. So I fixed the problem that way.
